### ISSUE TITLE
Base URL is wrong with 3.0.0.SNAPSHOT

### ISSUE THREAD

**Issue by ogerardin** [springfox/springfox#3030]
I have a Spring Boot project deployed with server.servlet.contextPath=/data

With Springfox 2.9.2, the behaviour of swagger-ui is as expected: base URL is /data and operation URLs are relative to the base URL, e.g. /version

If I just replace with Springfox 3.0.0-SNAPSHOT (changing @EnableSwagger2 to @EnableSwagger2WebMvc) then base URL is displayed as "/" in swagger-ui.html, and the base URL is now included in the operation URLs, e.g. /data/version

Accessing the endpoints from swagger-ui works since the computed actual URL is the same, but the generated swagger file is useless if you deploy to another base URL.

What version of the library are you using?
- [X] 3.0.0-SNAPSHOT

What kind of issue is this?
- [X] Bug report

**Comment by ogerardin** [comment]
It would appear that the base path is hardcoded to ROOT = "/" which means that there is no way to override it. I can only assume that it's a temporary hack.

```
DocumentationBuilder group = new DocumentationBuilder()
        .name(context.getGroupName())
        .apiListingsByResourceGroupName(apiListings)
        .produces(context.getProduces())
        .consumes(context.getConsumes())
        .host(context.getHost())
        .schemes(context.getProtocols())
        .basePath(ROOT)
        .extensions(context.getVendorExtentions())
        .tags(tags);
```

Compare with the 2.9.2 version:
```
DocumentationBuilder group = new DocumentationBuilder()
        .name(context.getGroupName())
        .apiListingsByResourceGroupName(apiListings)
        .produces(context.getProduces())
        .consumes(context.getConsumes())
        .host(context.getHost())
        .schemes(context.getProtocols())
        .basePath(context.getPathProvider().getApplicationBasePath())
        .extensions(context.getVendorExtentions())
        .tags(tags);
```

**Comment by ogerardin** [comment]
The latest snapshot (20190804) now has the base URL correct, but unfortunately the context base is still also part of each endpoint's URL... As a consequence calling endpoints form Swagger-ui doesn't work at all if there's a context path other than /.

For me the situation is worse, because at least with the previous version the effective URL was correct, so swagger-ui was still usable (even though the generated swagger file was wrong), but now the effective URL has the context base doubled...

To summarize, assuming contextPath=/data:

- v2.9.2
  * base URL: /data :white_check_mark:
  * endpoints URL: /xxx :white_check_mark:
  * effective URL: /data/xxx :white_check_mark:

- previous 3.0.0-SNAPSHOT
  * base URL: / :x:
  * endpoints URL: /data/xxx :x:
  * effective URL: /data/xxx :white_check_mark:

- current 3.0.0-SNAPSHOT (20190804)
  * base URL: /data :white_check_mark:
  * endpoints URL: /data/xxx :x:
  * effective URL: /data/data/xxx :x:

**Comment by dilipkrish** [comment]
Im interested in understanding issue more. How are you overriding the base path? Are you using an API gateway?

**Comment by ogerardin** [comment]
> Im interested in understanding issue more. How are you overriding the base path? Are you using an API gateway?

Nothing fancy really, just the Spring property server.servlet.contextPath=/data
All 3 tests were done with the same code base and configuration, just switching the Springfox version.

**Comment by ogerardin** [comment]
No update on this topic ?

**Comment by ogerardin** [comment]
OK here's a workaround for this issue: using a custom PathProvider I can remove the duplicate base path.

 ```
@Value("${server.servlet.contextPath}")
    private String contextPath;
```

  ```
Docket docket = new Docket(DocumentationType.SWAGGER_2)
                ...
                .pathProvider(new MyPathProvider())
                ...

```
 ```
private class MyPathProvider extends DefaultPathProvider {
        @Override
        public String getOperationPath(String operationPath) {
            if (operationPath.startsWith(contextPath)) {
                operationPath = operationPath.substring(contextPath.length());
            }
            return removeAdjacentForwardSlashes(UriComponentsBuilder.newInstance().replacePath(operationPath)
                    .build().toString());
        }
    }
```

**Comment by JonBainbridge** [comment]
This is still an issue when running Spring-Boot 2.2.2.RELEASE and SpringFox 3.0.0-SNAPSHOT.

However slight changes to the syntax for the workaround above.

@Value("${server.servlet.context-path}")
	private String contextPath;

Paths.removeAdjacentForwardSlashes(.....);

**Comment by ogerardin** [comment]
Forgot to mention, using Spring boot 2.1.6 here.

**Comment by kokhanchyk** [comment]
as a workaroud for this issue you can use 

`
@Configuration
@Profile("swagger")
@EnableSwagger2WebFlux
public class SwaggerConfiguration {

private static final String SWAGGER_API_DOCS = "/v2/api-docs";

  @Value("${springfox.documentation.swagger.v2.path}")
  private String swagger2Endpoint;

@Bean
  RouterFunction<ServerResponse> proxyRouterForSwagger() {
    return route(GET(swagger2Endpoint), req -> forwardRequest(SWAGGER_API_DOCS));
  }

  private Mono<ServerResponse> forwardRequest(String destinationUrl) {
    return ok().body(WebClient
           .create("http://localhost:8080")
           .get()
           .uri(destinationUrl)
           .retrieve()
           .bodyToMono(JsonNode.class), JsonNode.class);
  }

  @Bean
  public Docket api(final TypeResolver typeResolver) {
    return new Docket(DocumentationType.SWAGGER_2)
          .host("api.your.host.com")
          .select()
          .apis(RequestHandlerSelectors.any())
          .paths(PathSelectors.any())
          .build()
            .protocols(newHashSet("http", "https"))
            .securitySchemes(newArrayList(jwtScheme()))
             .securityContexts(newArrayList(new SecurityContext(jwtAuth(),
                     regex("/xxxxx*"))));
  }
}
`

When you set swagger2Endpoint, and will call it, it will forward to standard "/v2/api-docs" under the hood.
Also dont forget to specify host in Docket bean, because it will overide by localhost. 
And port in localhost in   WebClient.create("http://localhost:8080") should be execly the same as your service is running on.

**Comment by schneivo77** [comment]
Still an issue with Spring Boot 2.3.0.RELEASE and SpringFox 3.0.0-SNAPSHOT (latest version when I created this post).

**Comment by abhinavmishra14** [comment]
This issue seems to be with Spring Boot 2.3.1.RELEASE and SpringFox 3.0.0-SNAPSHOT as well. Is there any possibility that this bug could be fixed ?

**Comment by dilipkrish** [comment]
This should now be fixed in the latest snapshot

**Comment by abhinavmishra14** [comment]
> This should now be fixed in the latest snapshot

hi @dilipkrish 

Seems like swagger-ui is broken with latest sanpshot, when i use contextpath.  My context path is: "/rwsspringboot"

I tried both these URLs and none works:
http://127.0.0.1:8181/rwsspringboot/swagger-ui/index.html
http://127.0.0.1:8181/rwsspringboot/swagger-ui.html

`<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-swagger2</artifactId>
			<version>3.0.0-SNAPSHOT</version>
		</dependency>

		<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-swagger-ui</artifactId>
			<version>3.0.0-SNAPSHOT</version>
		</dependency>
		
		<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-spring-webmvc</artifactId>
			<version>3.0.0-SNAPSHOT</version>
		</dependency>`

**Comment by abhinavmishra14** [comment]
[Update:] With latest snapshot, even swagger-ui stopped working without context path.

None works:

http://127.0.0.1:8181/swagger-ui
http://127.0.0.1:8181/swagger-ui.html
http://127.0.0.1:8181/swagger-ui/index.html

**Comment by dilipkrish** [comment]
See [here](https://github.com/springfox/springfox/issues/3070#issuecomment-650591154) and please dont crosspost.

**Comment by dilipkrish** [comment]
Is it any better now @ogerardin?

**Comment by ogerardin** [comment]
Nope.. We're back to this situation:

> base URL: / ❌
> endpoints URL: /data/xxx ❌
> effective URL: /data/xxx ✅
> 

Which means you can call call the endpoints using Swagger-UI's "Try it" because the effective URL is correct, but the base URL is wrong (contextPath is ignored) so the swagger file can't be used if you deploy on another context path....

Worse, my workaround with a custom PathProvider doesn't work anymore, as it removes the duplicated base URL from each operation's URL, but since the base URL is wrong, the effective URL becomes wrong :(

So actually it's worse than before for me.

**Comment by dilipkrish** [comment]
Have you tried using OAS_30 rather than SWAGGER_2? For your documentation type?

**Comment by ogerardin** [comment]
I just did, and it's exactly the same: base URL is / when it should contain the context path, and the context path is repeated in each operation's URL.

**Comment by dilipkrish** [comment]
OpenAPI doesn't have any concept of base path. It only deals with servers, it is designed to be portable, can you show me what you mean? ... for e.g. here is the [current openAPI spec](https://github.com/springfox/springfox/blob/master/oas-contract-tests/src/test/resources/contracts/features.json)

**Comment by ogerardin** [comment]
> OpenAPI doesn't have any concept of base path.

It has the concept of base URL, see https://swagger.io/docs/specification/api-host-and-base-path/
This means that in the case of Spring Boot, the context path should be part of the server URL, as shown in the examples.

In my case I expect the server.url to be "http://localhost:8080/data", and that "/data" is not part of each endpoint's path.

**Comment by dilipkrish** [comment]
Well since its impossible to figure out all the different use cases, I've provided an existensibilty mechanism. Use it wisely 😉

**Comment by ogerardin** [comment]
Do you have an example on how to use it to do that (include context path in base URL) ?

I still believe that this should be the default when using Spring boot with a context path. Why on earth would anyone want to include the context path in each endpoint path? The controllers only define the endpoint path; the context path is where you graft them. The endpoints path must not depend on the context path.

**Comment by dilipkrish** [comment]
Here is [an example](https://github.com/springfox/springfox/blob/4f4aa545e7716f9d5a8dae2d855a6205d403b0b5/springfox-oas/src/main/java/springfox/documentation/oas/web/WebMvcBasePathAndHostnameTransformationFilter.java#L20). Im trying to get the release out. It works for the 80% or use cases and possibly doesn't work for applications in AWS, but the idea is to not wait to get this perfect, but to get it released so that we can get feedback and improve it in the subsequent release. 

Since you're a little bit more familiar with this, would love your input, perhaps even a PR to fix the default transformation filters.

**Comment by ogerardin** [comment]
I'm having trouble figuring out how to use WebMvcOpenApiTransformationFilter. Is it just for OpenAPI or also Swagger 2.0?
When exactly is this transformation called? How do I use it to do what I want, i.e. move the context path from the endpoint paths to the base URL? If I add my own, will it not conflict with WebMvcBasePathAndHostnameTransformationFilter?

**Comment by ogerardin** [comment]
OK so this is only for OpenAPI, the Swagger 2 version is at springfox.documentation.swagger2.web.WebMvcBasePathAndHostnameTransformationFilter

I have **almost** managed to do what I want with this code:
```
  public Swagger transform(SwaggerTransformationContext<HttpServletRequest> context) {
    Swagger swagger = context.getSpecification();
    context.request().ifPresent(servletRequest -> {
      UriComponents uriComponents = componentsFrom(servletRequest, swagger.getBasePath());
      swagger.host(String.format("%s:%d", uriComponents.getHost(), uriComponents.getPort()));
      String basePath = isEmpty(uriComponents.getPath()) ? "/" : uriComponents.getPath();
      swagger.basePath(basePath);
      final Map<String, Path> newPaths = swagger.getPaths().entrySet().stream()
              .collect(Collectors.toMap(
                      entry -> entry.getKey().replaceAll("^" + basePath, ""),
                      Map.Entry::getValue));
      swagger.setPaths(newPaths);
    });
    return swagger;
  }

```
Unfortunately the default WebMvcBasePathAndHostnameTransformationFilter runs after mine and overwrites the base path. How do I disabled the default one?

Also, I feel it's a very awkward way to do things. The base path and endpoint paths should be correct in the Swagger object in the first place.

**Comment by dilipkrish** [comment]
Like any other plugin its ordered. If you add an annotation to your component `@Order(Ordered.HIGHEST_PRECEDENCE + 1000) it should do the trick. Getting the basePath can be tricky because of the variance of how services can be set up behind load balancers. This is the approach that was quickest to the release. Happy to incorporate any suggestion that you might have as long as it works for 100% of the cases.

**Comment by ogerardin** [comment]
Plugin ordering is confusing, increasing the precedence actually makes it run later! Anyway it now works, thanks.

I stand to my statement: in a typical Spring Boot app, with a context path, accessed locally, the context path should be part of the Swagger base path / OpenAPI base URL and not part of each endpoint, which isn't the case now. If I had time I would investigate how to do that without involving plugins, but unfortunately I don't :(

**Comment by dilipkrish** [comment]
I don't either. That's why we use plugins so that if you're not happy with the framework decisions you are empowered to fix it for your specific use case. When I have time to fix it or if i find a more generic solution that applies broadly I'm happy to incorporate it into the base framework. Hope you understand.

Thanks for pressing on this issue. I'm hoping you or someoneelse who reads this might have a great idea before I do. Cheers!

**Comment by ogerardin** [comment]
For the record here's the complete workaround class I use. It's only for Swagger 2 but I'm sure you can write a custom springfox.documentation.oas.web.WebMvcBasePathAndHostnameTransformationFilter for OpenAPI in the same way.

```
@Component
// make sure it runs after the default plugin springfox.documentation.swagger2.web.WebMvcBasePathAndHostnameTransformationFilter
@Order(Ordered.HIGHEST_PRECEDENCE + 1000)
public class CustomWebMvcBasePathAndHostnameTransformationFilter implements WebMvcSwaggerTransformationFilter {
  private final Environment env;

  public CustomWebMvcBasePathAndHostnameTransformationFilter(Environment env) {
    this.env = env;
  }

  @Override
  public Swagger transform(SwaggerTransformationContext<HttpServletRequest> context) {
    Swagger swagger = context.getSpecification();
    context.request().ifPresent(servletRequest -> {
      UriComponents uriComponents = componentsFrom(servletRequest, swagger.getBasePath());
      swagger.host(String.format("%s:%d", uriComponents.getHost(), uriComponents.getPort()));
      // set Swagger base path
      String basePath = isEmpty(uriComponents.getPath()) ? "/" : uriComponents.getPath();
      swagger.basePath(basePath);
      // rewrite paths to remove the leading base path from each endpoint
      final Map<String, Path> newPaths = swagger.getPaths().entrySet().stream()
              .collect(Collectors.toMap(
                      entry -> entry.getKey().replaceAll("^" + basePath, ""),
                      Map.Entry::getValue));
      swagger.setPaths(newPaths);
    });
    return swagger;
  }

  @Override
  public boolean supports(DocumentationType delimiter) {
    return delimiter == DocumentationType.SWAGGER_2;
  }
}

```

**Comment by dilipkrish** [comment]
Awesome thank you for sharing that. I'll see if I can incorporate this without breaking swagger-ui

**Comment by martyncrow** [comment]
I agree with @ogerardin it makes sense that the context path is set against the basePath (swagger 2.0) or baseUrl (OpenAPI). Having the contextPath prepended to every path in the spec causes issues if you are trying to use the spec to update an API Gateway or simillar.

**Comment by prokhorovalexey** [comment]
I agree with @martyncrow . Another point here is to do not break previous behavior and so let us upgrade with less pain.

**Comment by ogerardin** [comment]
> Having the contextPath prepended to every path in the spec causes issues if you are trying to use the spec to update an API Gateway or simillar. 

That's exactly my use case, the swagger file is used to update an API definition in Azure's API Management. Without my workaround the swagger file is tied to the contextPath of the specific machine it was generated on.

**Comment by sgmccreary** [comment]
@ogerardin I'd like to try the workaround that you implemented.  Can you share the details of the componentsFrom method and Path object?

**Comment by prokhorovalexey** [comment]
A bit modified plugin originally posted by @ogerardin ogerardin:
`package com.elc.d2c.interceptor;

import io.swagger.models.Path;
import io.swagger.models.Swagger;
import org.apache.commons.lang3.StringUtils;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;
import org.springframework.web.util.UriComponents;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.swagger.common.HostNameProvider;
import springfox.documentation.swagger2.web.SwaggerTransformationContext;
import springfox.documentation.swagger2.web.WebMvcSwaggerTransformationFilter;

import javax.servlet.http.HttpServletRequest;
import java.util.Map;
import java.util.stream.Collectors;

@Component
// make sure it runs after the default plugin springfox.documentation.swagger2.web.WebMvcBasePathAndHostnameTransformationFilter
@Order(Ordered.HIGHEST_PRECEDENCE + 1000)
public class CustomWebMvcBasePathAndHostnameTransformationFilter implements WebMvcSwaggerTransformationFilter {

    @Override
    public Swagger transform(SwaggerTransformationContext<HttpServletRequest> context) {
        Swagger swagger = context.getSpecification();
        context.request().ifPresent(servletRequest -> {
            UriComponents uriComponents = HostNameProvider.componentsFrom(servletRequest, swagger.getBasePath());

            swagger.host(uriComponents.getHost() + (uriComponents.getPort() > 0 ? ":" + uriComponents.getPort() : ""));

            // set Swagger base path
            String basePath = StringUtils.isEmpty(uriComponents.getPath()) ? "/" : uriComponents.getPath();
            swagger.basePath(basePath);

            // rewrite paths to remove the leading base path from each endpoint
            final Map<String, Path> newPaths = swagger.getPaths().entrySet().stream().collect(Collectors.toMap(
                    entry -> entry.getKey().replaceAll("^" + basePath, ""), Map.Entry::getValue));
            swagger.setPaths(newPaths);
        });
        return swagger;
    }

    @Override
    public boolean supports(DocumentationType delimiter) {
        return delimiter == DocumentationType.SWAGGER_2;
    }

}`

**Comment by dilipkrish** [comment]
Thank you @ogerardin and @prokhorovalexey! Does this work as a generic solution? as in does in work in an ALB/ELB scenario as well as on a local machine? If so would love to fold this back in to the 3.0.1 release.

**Comment by prokhorovalexey** [comment]
Yes, it works for me via standard Spring component annotation.

**Comment by dilipkrish** [comment]
By work I mean does swagger ui work as expected under both scenarios? And are u able to use the "Try it out" to exercise the api?

**Comment by prokhorovalexey** [comment]
Yes.

**Comment by dilipkrish** [comment]
Nice! Let me port this over to the library then... or would you like to try that?

**Comment by ogerardin** [comment]
@prokhorovalexey 
> A bit modified plugin originally posted by @ogerardin:

Thanks, I had made the modification for when port is -1, but I forgot to post the update ! 

@sgmccreary you can use @prokhorovalexey 's code, it has all the imports :)

**Comment by ogerardin** [comment]
The pathProvider trick doesn't work anymore with 3.0.0, so you're left with this ugly workaround, just like the rest of us...
I really wish there was a simple way to control the base URL...

**Comment by dilipkrish** [comment]
This will be addressed in 3.0.1 @ogerardin

**Comment by ogerardin** [comment]
There is a bug in the previous solution, it only works when the base path is not empty... If it's empty it strips the leading slash from all paths.

Here's a fixed version that works in both cases:

```java
    @Override
    public Swagger transform(SwaggerTransformationContext<HttpServletRequest> context) {
        Swagger swagger = context.getSpecification();
        context.request().ifPresent(servletRequest -> {
            UriComponents uriComponents = componentsFrom(servletRequest, swagger.getBasePath());
            swagger.host(hostName(uriComponents));

            String basePath;
            if (!isEmpty(uriComponents.getPath())) {
                basePath = uriComponents.getPath();
                // rewrite paths to remove the leading base path from each endpoint
                final Map<String, Path> newPaths = swagger.getPaths().entrySet().stream()
                        .collect(Collectors.toMap(
                                entry -> entry.getKey().replaceAll("^" + basePath, ""),
                                Map.Entry::getValue));
                swagger.setPaths(newPaths);
            } else {
                basePath = "/";
                // no need to rewrite paths
            }
            swagger.basePath(basePath);
        });
        return swagger;
    }

    private String hostName(UriComponents uriComponents) {
        String host = uriComponents.getHost();
        int port = uriComponents.getPort();
        if (port > -1) {
            return String.format("%s:%d", host, port);
        }
        return host;
    }
```

**Comment by markhu** [comment]
Found a simpler workaround:

```java
  @Value("${spring.profiles.active:}")
  private String activeSpringProfile;

...
        return new Docket(DocumentationType.SWAGGER_2)
...
            .pathMapping(activeSpringProfile.contains("local") ? "" : "/myPrefix")
...
```

Oddly enough, it doesn't set Swagger's `basePath` but it seems to end up with equivalent results.

P.S. So similar to this older issue: https://github.com/springfox/springfox/issues/2817

**Comment by mad** [comment]
Modified version for OpenAPI

```java
import java.util.*;
import javax.servlet.http.HttpServletRequest;

import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.Paths;
import io.swagger.v3.oas.models.servers.Server;
import springfox.documentation.oas.web.OpenApiTransformationContext;
import springfox.documentation.oas.web.WebMvcOpenApiTransformationFilter;
import springfox.documentation.spi.DocumentationType;

@Component
@Order(Ordered.HIGHEST_PRECEDENCE + 1000)
public class CustomWebMvcOpenApiTransformationFilter implements WebMvcOpenApiTransformationFilter {
    @Override
    public OpenAPI transform(OpenApiTransformationContext<HttpServletRequest> context) {
        OpenAPI openApi = context.getSpecification();
        context.request().ifPresent(servletRequest -> {
            String basePath = servletRequest.getContextPath();
            final Server server = openApi.getServers().get(0);
            openApi.setServers(Collections.singletonList(new Server().url(server.getUrl() + basePath)
                                                                     .description(server.getDescription())));
            final Paths paths = new Paths();
            openApi.getPaths().forEach((key, value) -> paths.addPathItem(key.replaceAll(basePath, ""), value));
            openApi.setPaths(paths);
        });
        return openApi;
    }

    @Override
    public boolean supports(DocumentationType delimiter) {
        return delimiter == DocumentationType.OAS_30;
    }
}

```

**Comment by marcellodesales** [comment]
# 👽 Netflix Zull Swagger API Docs for Microservices

* We have a requirement to expose the Swagger APIs internally from an internal Gateway as a consolidated view of the APIs
  * API Gateway - `Netflix Zuul` 
* The URLs returned were from the Gateway instead of the endpoint itself
  * Solution was to use the proposed component above from 

<img width="1329" alt="Screen Shot 2021-09-15 at 9 36 46 PM" src="https://user-images.githubusercontent.com/131457/133550378-46fe3807-eabd-489a-9b6e-48cc81ff1363.png">


## ☕ Swagger 2.0 API with Microservice Endpoint instead of Gateway

* Just verifying if the requested API docs were from a Gateway/Proxy
  * Anything with a request with `x-forwarded-for`

```java
import com.google.common.base.Strings;
import io.swagger.models.Path;
import io.swagger.models.Swagger;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.util.UriComponents;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.swagger2.web.SwaggerTransformationContext;
import springfox.documentation.swagger2.web.WebMvcSwaggerTransformationFilter;

import javax.servlet.http.HttpServletRequest;
import java.util.Collections;
import java.util.Map;
import java.util.stream.Collectors;

import springfox.documentation.swagger.common.HostNameProvider;

/**
 * Changes the host and path to make sure it returns the proper when returned by the gateway
 */
// make sure it runs after the default plugin springfox.documentation.swagger2.web.WebMvcBasePathAndHostnameTransformationFilter
// Based on https://github.com/springfox/springfox/issues/3030#issuecomment-662982083
@Component
@Order(Ordered.HIGHEST_PRECEDENCE + 1000)
public class CustomWebMvcBasePathAndHostnameTransformationFilter implements WebMvcSwaggerTransformationFilter {

    protected static final Logger LOG = LoggerFactory.getLogger(CustomWebMvcBasePathAndHostnameTransformationFilter.class);

    @Override
    public boolean supports(DocumentationType delimiter) {
        return delimiter == DocumentationType.SWAGGER_2;
    }

    @Override
    public Swagger transform(SwaggerTransformationContext<HttpServletRequest> context) {
        Swagger swagger = context.getSpecification();
        context.request().ifPresent(servletRequest -> {
            // Map the request headers to their values
            // https://stackoverflow.com/questions/23261803/iterate-an-enumeration-in-java-8/23261952#23261952
            // https://stackoverflow.com/questions/21015889/why-does-getheadernames-returns-an-enumeration-in-httpservletrequest-but-a-colle
            Map<String, String> requestHeaderIndex = Collections.list(servletRequest.getHeaderNames()).stream()
                    .collect(Collectors.toMap(headerName -> headerName, headerName -> servletRequest.getHeader(headerName.toString())));

            UriComponents uriComponents = HostNameProvider.componentsFrom(servletRequest, swagger.getBasePath());

            LOG.debug("Requested with the following headers: {}", requestHeaderIndex);
            // If this was a call from the gateway, use the URL from the current host instead of the server
            swagger.host(requestedHost(uriComponents, servletRequest));

            String basePath;
            if (!StringUtils.isEmpty(uriComponents.getPath())) {
                basePath = uriComponents.getPath();
                // rewrite paths to remove the leading base path from each endpoint
                final Map<String, Path> newPaths = swagger.getPaths().entrySet().stream()
                        .collect(Collectors.toMap(
                                entry -> entry.getKey().replaceAll("^" + basePath, ""),
                                Map.Entry::getValue));
                swagger.setPaths(newPaths);
            } else {
                basePath = "/";
                // no need to rewrite paths
            }
            swagger.basePath(basePath);
        });
        return swagger;
    }

    /**
     * @param uriComponents the uri components
     * @return the host URL when calls goes directly to the server without any proxy just return the component
     */
    private String requestedHost(UriComponents uriComponents, HttpServletRequest request) {
        // When using an API Gateway, don't return the URI from the gateway, just from the server if serving the UI
        boolean requestedForwardedByProxy = !Strings.isNullOrEmpty(request.getHeader("x-forwarded-host"));
        String host = requestedForwardedByProxy ? request.getRemoteHost() : uriComponents.getHost();
        int port = requestedForwardedByProxy ? request.getServerPort() : uriComponents.getPort();
        if (port > -1) {
            return String.format("%s:%d", host, port);
        }
        return host;
    }
}
```

# 🔉 Testing Logs 

## Request from API Gateway

* It will use the value from the `servletRequest.getRemoteHost()` since the gateway forwarded the call to the microservice
  * In this example, `localhost:8500` Swagger UI loaded the Microservice's Swagger at `localhost:8503`

```console
$ curl -i localhost:8500/auth/swagger/docs/v2
HTTP/1.1 200 
X-Supercash-Gateway-Build-Version: --env-var-UNMAZEDBOOT_BUILDER_GIT_SHA-not-provided-----env-var-UNMAZEDBOOT_BUILDER_GIT_BRANCH-not-provided--
X-Supercash-Gateway-Api-Version: v2
Vary: Origin
Vary: Access-Control-Request-Method
Vary: Access-Control-Request-Headers
X-Supercash-Build-Version: --env-var-UNMAZEDBOOT_BUILDER_GIT_SHA-not-provided-----env-var-UNMAZEDBOOT_BUILDER_GIT_BRANCH-not-provided--
X-Supercash-Api-Version: v2
X-B3-TraceId: 0e4522483333a072
Date: Thu, 16 Sep 2021 04:31:21 GMT
Keep-Alive: timeout=60
Content-Type: application/json
Transfer-Encoding: chunked
```
```json
{"swagger":"2.0","info":{
"host":"127.0.0.1:8503","basePath":"/","tags":
...
}}}%                                                                                                                                                                      
```

## Directly to the service 

* It will use the regular, it goes as before with the component itself that carries the value of the host and port as usual

```console
$ curl -i localhost:8503/swagger/docs/v2     
HTTP/1.1 200 
X-Supercash-Build-Version: --env-var-UNMAZEDBOOT_BUILDER_GIT_SHA-not-provided-----env-var-UNMAZEDBOOT_BUILDER_GIT_BRANCH-not-provided--
X-Supercash-Api-Version: v2
X-B3-TraceId: 35454bc4b13eb273
Content-Type: application/json
Transfer-Encoding: chunked
Date: Thu, 16 Sep 2021 04:31:26 GMT

{"swagger":"2.0","info":{"host":"localhost:8503","basePath":"/","tags":
...
}}}
```
